{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "from pprint import pprint\n",
    "import json\n",
    "import pickle\n",
    "import re\n",
    "from collections import OrderedDict\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from rdkit import Chem as chem\n",
    "from rdkit.Chem import AllChem, Draw\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "sys.path.insert(0, '/clusterCAD')\n",
    "os.environ.setdefault(\"DJANGO_SETTINGS_MODULE\", \"clusterCAD.settings\")\n",
    "import django\n",
    "django.setup()\n",
    "import pks.models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1396 MiBiG entries found!\n"
     ]
    }
   ],
   "source": [
    "file_path = './mibig'\n",
    "file_names = glob.glob(os.path.join(file_path, '*.json'))\n",
    "print('%d MiBiG entries found!' %(len(file_names)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Type III', 'Modular type I', 'Iterative typeI', 'Trans-AT type I', 'Modular Type I', 'Iterative type I', 'Type II', 'Other', 'PUFA synthase or related', 'Enediyne type I', 'Type I'}\n"
     ]
    }
   ],
   "source": [
    "# Print PKS subtype labels there present\n",
    "labels = []\n",
    "for file_name in file_names:\n",
    "    with open(file_name) as json_file:\n",
    "        json_data = json.load(json_file)\n",
    "    try:\n",
    "        labels.extend(json_data['general_params']['Polyketide']['pks_subclass'])\n",
    "    except KeyError:\n",
    "        pass\n",
    "print(set(labels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "352 potential type I modular PKS clusters found!\n"
     ]
    }
   ],
   "source": [
    "# This gets all the modular type I PKSs contained in MiBiG\n",
    "t1pks = []\n",
    "for file_name in file_names:\n",
    "    with open(file_name) as json_file:\n",
    "        json_data = json.load(json_file)\n",
    "    try:        \n",
    "        if len(set(['Modular type I', 'Modular Type I', 'Type I']).intersection(set(json_data['general_params']['Polyketide']['pks_subclass']))) > 0:\n",
    "            accession = json_data['general_params']['loci']['nucl_acc'][0]['Accession']\n",
    "            t1pks.append((file_name.split('/')[-1].split('.')[0], accession))\n",
    "    except KeyError:\n",
    "        pass\n",
    "# Number of clusters found\n",
    "print('%d potential type I modular PKS clusters found!' %(len(t1pks)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# antiSMASH output\n",
    "antismash_file_path = './mibig/antismash/split_files'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Proccessing functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def process_subunit_modules(sec_met): \n",
    "    '''This function takes as input annotated features for a PKS subunit (stored  as feature.qualifiers['sec_met'] \n",
    "        after annotation by antiSMASH), and returns a OrderedDict of OrderedDict objects corresponding to the \n",
    "        annotated modules contained by the subunit. They keys of the subunit OrderedDict are the index of the \n",
    "        module within the subunit. This function assumes that feature.type=='CDS' and that feature.qualifiers \n",
    "        has the key 'sec_met'. If the first entry of 'sec_met' is not 'Type: t1pks' then nothing is returned.\n",
    "    '''\n",
    "    # Initialize dict for the subunit\n",
    "    # keys: module number\n",
    "    # values: OrderedDict of domains in module. within the OrderedDict, the key is the domain main, while the\n",
    "    #         value is a length two list where the first element is a dictionary {start:, stop:} and the \n",
    "    #         second element is specificity dictionary \n",
    "    subunit = {}\n",
    "    \n",
    "    # This is for the current module (function processes subunit which may have more than one module)\n",
    "    module_index = 0  # key for module\n",
    "    module_domains = [] # list of domains in module\n",
    "    old_module_domains = [] # pre-initialize in case subunit starts with a domain\n",
    "                            # that is expected to end the module\n",
    "    \n",
    "    # This is how domains appear in sec_met:\n",
    "    # ['PKS_AT', 'PKS_KS', 'PKS_KR', 'PKS_DH', 'PKS_ER', 'ACP', 'Thioesterase']\n",
    "    # Iterate over the entries in sec_met, and add them to the module_domains list \n",
    "    for entry in sec_met:    \n",
    "        # Split entry into a list\n",
    "        entrysplit = [item.strip() for item in entry.split(';') if item != '']\n",
    "        # Split part of entry that is expected to describe catalytic domain\n",
    "        domainsplit = entrysplit[0].split()\n",
    "        # This is just different ways of processing the name of the domain depending\n",
    "        # on how the name of the domain is formatted\n",
    "        if ' '.join(domainsplit[:2]) == 'NRPS/PKS Domain:' and len(domainsplit) > 2:\n",
    "            # Note that we want to make sure that there is a leading 'PKS_' before we do our trimming\n",
    "            if domainsplit[2].split('_')[0] == 'PKS':\n",
    "                if domainsplit[2] in ['PKS_Docking_Nterm', 'PKS_Docking_Cterm']:\n",
    "                    domaintype = domainsplit[2]\n",
    "                else:\n",
    "                    # Trim off the leading 'PKS_'\n",
    "                    # Assume 'DH2' and 'DHt' are the same as 'DH' \n",
    "                    domaintype = domainsplit[2].split('_')[-1].replace('DHt', 'DH').replace('DH2', 'DH')\n",
    "            # Special case of 'CAL' domain\n",
    "            elif domainsplit[2] == 'CAL_domain':\n",
    "                domaintype = 'CAL'\n",
    "            else:\n",
    "                domaintype = domainsplit[2]\n",
    "        else:\n",
    "            continue\n",
    "        \n",
    "        # These are the catalytic domains that we want to recognize\n",
    "        if domaintype not in ['KS', 'AT', 'KR', 'DH', 'ER', 'ACP', 'Thioesterase', \n",
    "                              'cMT', 'oMT', 'CAL', 'PCP', \n",
    "                              'Heterocyclization', 'AMP-binding', \n",
    "                              'Condensation_DCL', 'Condensation_LCL',\n",
    "                              'PKS_Docking_Nterm', 'PKS_Docking_Cterm']:\n",
    "            # Break out of for loop and stop looking for additional catalytic domains if \n",
    "            # we encounter a domain that we don't recognize\n",
    "            # We end up excluding any subunit that has a non-recognized catalytic domain\n",
    "            break    \n",
    "        # Get the boundaries of the catalytic domain\n",
    "        boundaries = [int(bound) for bound in domainsplit[3].replace('(', '').replace(')', '').replace('.', '').split('-')]\n",
    "        \n",
    "        # Here, we add each domain to a list, which will be converted to an OrderedDict\n",
    "        # based on whether or not the domain is expected to have substrate specificity annotations\n",
    "        if domaintype in ['KS', 'DH', 'ER', 'ACP', 'cMT', 'oMT', 'CAL', 'PCP',\n",
    "                          'Heterocylization', 'AMP-binding', \n",
    "                          'Condensation_DCL', 'Condensation_LCL',\n",
    "                          'PKS_Docking_Nterm', 'PKS_Docking_Cterm']:\n",
    "            module_domains.append((domaintype, [{'start': boundaries[0], 'stop': boundaries[1]}]))\n",
    "        # Include substrate and stereospecificity annotations for AT and KR domains respectively\n",
    "        if domaintype in ['AT', 'KR']:\n",
    "            notesdict = {}\n",
    "            for note in entrysplit[1:]:\n",
    "                item = note.split(': ')\n",
    "                notesdict[item[0]] = item[1]\n",
    "            module_domains.append((domaintype, [{'start': boundaries[0], 'stop': boundaries[1]}, notesdict]))\n",
    "                \n",
    "        # End of the module has been reached of the domain is 'ACP' or 'PCP\n",
    "        if domaintype in ['ACP', 'PCP']:\n",
    "            domains_present = [d[0] for d in module_domains]\n",
    "            # Make sure every module has an AT, or else it isn't a valid module and we just ignore it\n",
    "            # This means it will be excluded from the subunit, which makes sense since we can't \n",
    "            # really perform a polyketide chain extension without an AT\n",
    "            if 'AT' in domains_present:            \n",
    "                subunit[module_index] = OrderedDict(module_domains)\n",
    "                old_module_domains = module_domains\n",
    "                module_index += 1\n",
    "            else:\n",
    "                old_module_domains = []\n",
    "            module_domains = []\n",
    "        # These domains may come after the ACP or PCP, so if they are encountered, we add\n",
    "        # them to previous module and keep going forward\n",
    "        if domaintype in ['Thioesterase', 'PKS_Docking_Cterm', 'Condensation_LCL']:\n",
    "            # Overwrite previous subunit, or else will have duplicate entries\n",
    "            old_module_domains.append((domaintype, [{'start': boundaries[0], 'stop': boundaries[1]}]))\n",
    "            subunit[module_index - 1] = OrderedDict(old_module_domains)\n",
    "            module_domains = []\n",
    "            \n",
    "    return subunit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_gene_data(record):\n",
    "    '''Takes as input a record read in from an MiBiG GenBank file using SeqIO.read() and outputs list of sublists\n",
    "        containing PKS data from that record. Each entry in list is a sublist that represents a subunit or standalone.\n",
    "    '''\n",
    "    # Get list to hold information about all genes that are in the record\n",
    "    gene_data = []\n",
    "    \n",
    "    # Only the \"CDS\" features are potentially gene\n",
    "    for feature in record.features:\n",
    "        # These are the features we are interested in\n",
    "        if feature.type == 'CDS' and 'protein_id' in feature.qualifiers.keys() and 'gene' in feature.qualifiers.keys(): \n",
    "            # This gets the location of the feature\n",
    "            location = feature.location\n",
    "            # General information about gene\n",
    "            if 'product' in feature.qualifiers.keys():\n",
    "                description = feature.qualifiers['product'][0]\n",
    "            gene_data.append([feature.qualifiers['protein_id'][0],\n",
    "                              feature.qualifiers['gene'][0],\n",
    "                             ])\n",
    "            # Feature may not be a PKS module and therefore may not have have subunits \n",
    "            # (this will be overwritten if it does have subunits)\n",
    "            subunit_modules = None\n",
    "            # Information if gene is PKS subunit\n",
    "            if 'sec_met' in feature.qualifiers.keys() and len(feature.qualifiers['sec_met']) > 3:\n",
    "                if feature.qualifiers['sec_met'][3] in ['NRPS/PKS subtype: Type I Modular PKS', \n",
    "                                                        'NRPS/PKS subtype: PKS-like protein',\n",
    "                                                        'NRPS/PKS subtype: PKS/NRPS-like protein',\n",
    "                                                        'NRPS/PKS subtype: Hybrid PKS-NRPS']:\n",
    "                    subunit_modules = process_subunit_modules(feature.qualifiers['sec_met'])\n",
    "\n",
    "            # Append description and position of gene within nucleotide sequence\n",
    "            gene_data[-1].extend([description, [location.start.position, location.end.position]])\n",
    "\n",
    "            # Subunit information (if it doesn't have subunit information, assumed to be a standalone enzyme)\n",
    "            if subunit_modules:\n",
    "                gene_data[-1].append(subunit_modules)\n",
    "\n",
    "            # Amino acid sequence\n",
    "            gene_data[-1].append(feature.qualifiers['translation'][0])\n",
    "\n",
    "    return gene_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def check_json_module_validity(module_list):\n",
    "    '''Function that makes sure module specified in JSON file is valid,\n",
    "       that is to say, make sure that it contains KS, AT, and ACP or PCP.\n",
    "       These domains have the following possible annotations in type I modular PKSs as of February 24, 2017:\n",
    "       ['KS', 'AT', 'T']\n",
    "       ['Ketosynthase', 'Acyltransferase', 'Thiolation (ACP/PCP)']\n",
    "    '''\n",
    "    at_check = len(set(['AT', 'Acyltransferase']).intersection(set(module_list)))\n",
    "    acp_check = len(set(['ACP', 'PCP', 'T', 'Thiolation (ACP/PCP)']).intersection(set(module_list)))\n",
    "\n",
    "    if at_check and acp_check:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def process_cluster(record, cluster_ref, mibig_json):\n",
    "    '''Takes in a record and corresponding MiBiG json file and enters cluster information into Django database.\n",
    "    '''\n",
    "    # Get information about the gene\n",
    "    gene_data = get_gene_data(record)\n",
    "    if len(gene_data) == 0:\n",
    "        return\n",
    "\n",
    "    # Initalize lists for subunits and standalones\n",
    "    # We make two dictionaries because sometimes the subunit name in the MiBiG JSON files\n",
    "    # is the gene name, e.g. eryA1, and sometimes it is the accession number, e.g. A0000000\n",
    "    unordered_subunits = {}\n",
    "    unordered_subunits_alt = {}\n",
    "    standalones = []\n",
    "     \n",
    "    # Recall that each entry in gene_data is a list\n",
    "    # [protein id, gene, product, [location start, location end], subunit dict (optional), translation]\n",
    "    \n",
    "    #####################\n",
    "    # Basic information #\n",
    "    #####################\n",
    "    \n",
    "    counter = 1\n",
    "    for gene in gene_data:\n",
    "        geneid = gene[0].strip()\n",
    "        genename = gene[1].strip()\n",
    "        genedesc = gene[2].strip()\n",
    "        genestart = gene[3][0]\n",
    "        genestop = gene[3][1]\n",
    "        genetranslation = gene[-1].strip()\n",
    "\n",
    "        # Just use length of gene_data to differentiate between standalones and subunits\n",
    "        if len(gene) == 6:\n",
    "            # We do this to take care of duplicated gene names, as is the case wity tylactone (BGC0000166)\n",
    "            if genename in unordered_subunits_alt.keys():\n",
    "                genename = genename + '_' + str(counter)\n",
    "                counter += 1\n",
    " \n",
    "            # Get subunit data from gene\n",
    "            genesubunitdata = gene[-2]\n",
    "            # Here we use the two dictionary options to save the unordered subunits\n",
    "            # Sometimes MiBiG uses geneid and sometimes it uses genename to reference subunits\n",
    "            unordered_subunits[geneid] = (genename, genedesc, genestart, genestop,\n",
    "                                            genesubunitdata, genetranslation)\n",
    "            unordered_subunits_alt[genename] = (geneid, genedesc, genestart, genestop,\n",
    "                                                genesubunitdata, genetranslation)\n",
    "        else:\n",
    "            # Standalones lack subunit and orphan entries\n",
    "            assert len(gene) == 5, gene\n",
    "            \n",
    "            #####################\n",
    "            # CREATE STANDALONE #\n",
    "            #####################\n",
    "\n",
    "#            pks.models.Standalone(cluster=cluster_ref)\n",
    "#            standalones.append(pks.Standalone(geneid, genename, genedesc, \n",
    "#                                              genestart, genestop, genetranslation))\n",
    "        \n",
    "    #########################################\n",
    "    # JSON file has cyclization information #\n",
    "    #########################################\n",
    "\n",
    "    # Get ordered version of subunits from corresponding JSON file\n",
    "    with open(mibig_json) as json_file:\n",
    "        mibig_data = json.load(json_file)\n",
    "    \n",
    "    # Get PKS cyclization information\n",
    "    # this will be either 'Cyclic' or 'Linear'\n",
    "    try:\n",
    "        lin_cycl_pk = mibig_data['general_params']['Polyketide']['lin_cycl_pk']\n",
    "        if lin_cycl_pk == 'Cyclic':\n",
    "            cyclize = True\n",
    "        elif lin_cycl_pk == 'Linear':\n",
    "            cyclize = False\n",
    "        else:\n",
    "            raise Exception(\"lin_cycl_pk expected to be 'Cyclic' or 'Linear'.\")\n",
    "    except KeyError:\n",
    "        cyclize = False\n",
    "            \n",
    "    #############################################\n",
    "    # Case 1: Use JSON file subunit information #\n",
    "    #############################################\n",
    "        \n",
    "    # Note that all gene data has now been processed, want to reprocess to get right ordering \n",
    "    # We strip out subunits that have invalid modules\n",
    "    try:\n",
    "        ordered_subunits = []\n",
    "        for subunit in mibig_data['general_params']['Polyketide']['mod_pks_genes']:\n",
    "            subunit_name = re.sub(r'\\s+', '', subunit['mod_pks_gene'])\n",
    "            subunit_modules = subunit['pks_module']\n",
    "\n",
    "            valid_subunit = True\n",
    "            # This checks if the module is valid\n",
    "            for module in subunit_modules:\n",
    "                if not check_json_module_validity(module['pks_domains']):\n",
    "                    valid_subunit = False\n",
    "            if valid_subunit:\n",
    "                ordered_subunits.extend(subunit_name.split(','))\n",
    "            else:\n",
    "                # Loop is broken once first invalid subunit is encountered\n",
    "                break\n",
    "        # If no valid subunits, then raise exception to use alphabetical ordering\n",
    "        if len(ordered_subunits) == 0:\n",
    "            raise Exception\n",
    "#             print('\\tNo valid subunits!')\n",
    "#             for subunit in mibig_data['general_params']['Polyketide']['mod_pks_genes']:\n",
    "#                 subunit_name = re.sub(r'\\s+', '', subunit['mod_pks_gene'])\n",
    "#                 subunit_modules = subunit['pks_module']\n",
    "#                 for module in subunit_modules:\n",
    "#                     print(module['pks_domains'])\n",
    "#             return\n",
    "        # This makes sure the subunit accession naming is consistent\n",
    "        # The purpose of these two 'if' statements is because there may be cases in the MiBiG JSON file\n",
    "        # where the name of the gene is for example, 'eryA1, A000000' and we want to keep consistant naming\n",
    "        if len(ordered_subunits[0]) >= 8:\n",
    "            ordered_subunits = [entry for entry in ordered_subunits if len(entry) >= 8]\n",
    "        if len(ordered_subunits) > 1:\n",
    "            if len(ordered_subunits[1]) >= 8:\n",
    "                ordered_subunits = [entry for entry in ordered_subunits if len(entry) >= 8]\n",
    "        # This is because sometimes the accession number under which the gene is recorded sometimes\n",
    "        # has a version number, and sometimes does not\n",
    "        if len(ordered_subunits[0].split('.')) == 1 and len(ordered_subunits[0]) == 8:\n",
    "            ordered_subunits = [entry + '.1' for entry in ordered_subunits]\n",
    "\n",
    "        # Check if subunit is in either dictionary\n",
    "        for isubunit,subunit in enumerate(ordered_subunits):\n",
    "            if subunit not in set(list(unordered_subunits.keys()) + list(unordered_subunits_alt.keys())):\n",
    "                print('Missing subunit: \"%s\"' %(subunit))\n",
    "                for gene in mibig_data['general_params']['Polyketide']['mod_pks_genes']:\n",
    "                    if gene['mod_pks_gene'] == subunit:\n",
    "                        module = gene['pks_module']\n",
    "                        for entry in module:\n",
    "                            print(entry['pks_domains'])\n",
    "                print(unordered_subunits.keys())\n",
    "                return\n",
    "    #    print([gene_ref for gene_ref in mibig_data['general_params']['Polyketide']['mod_pks_genes']])\n",
    "\n",
    "        # Determine whether to use standard or alternative dict\n",
    "        if len(ordered_subunits[0]) >= 8:\n",
    "            alt = False\n",
    "        else:\n",
    "            alt = True\n",
    "    \n",
    "    #########################################\n",
    "    # Case 2: Alphabetical subunit ordering #\n",
    "    #########################################\n",
    "    \n",
    "    # Just use unordered gene order if the gene ordering is not already in the JSON file \n",
    "    except Exception:\n",
    "        ordered_subunits = list(unordered_subunits_alt.keys())\n",
    "        ordered_subunits.sort()\n",
    "        alt = True\n",
    "\n",
    "    ####################################\n",
    "    # This does the subunit reordering #\n",
    "    ####################################\n",
    "    # This is just to make sure laoding module is assigned correctly\n",
    "    modules_seen = 0\n",
    "    for subunit_key in ordered_subunits:\n",
    "        # subunit data has form (id, description, start, stop, module dict, sequence)\n",
    "        if not alt:\n",
    "            subunitdata = unordered_subunits[subunit_key]\n",
    "        else:\n",
    "            subunitdata = unordered_subunits_alt[subunit_key]\n",
    "     \n",
    "        if not alt:\n",
    "            # subunit = id\n",
    "            # subunit[0] = name\n",
    "            # id, name, description, start, stop, sequence\n",
    "            subunit = pks.models.Subunit(cluster=cluster_ref,\n",
    "                                         genbankAccession=subunit_key,\n",
    "                                         name=subunitdata[0],\n",
    "                                         start=subunitdata[2],\n",
    "                                         stop=subunitdata[3],\n",
    "                                         sequence=subunitdata[-1])\n",
    "            subunit.save()\n",
    "            # subunits.append(pks.Subunit(subunit, subunitdata[0], subunitdata[1],\n",
    "            #                            subunitdata[2], subunitdata[3], subunitdata[-1],\n",
    "            #                            modules))\n",
    "        else:\n",
    "            # subunit = name\n",
    "            # subunit[0] = id\n",
    "            subunit = pks.models.Subunit(cluster=cluster_ref,\n",
    "                                         genbankAccession=subunitdata[0],\n",
    "                                         name=subunit_key,\n",
    "                                         start=subunitdata[2],\n",
    "                                         stop=subunitdata[3],\n",
    "                                         sequence=subunitdata[-1])\n",
    "            subunit.save()\n",
    "                \n",
    "                # subunits.append(pks.Subunit(subunitdata[0], subunit, subunitdata[1],\n",
    "                #                            subunitdata[2], subunitdata[3], subunitdata[-1],\n",
    "                #                            modules))\n",
    "        \n",
    "        # This is the modules for the subunit\n",
    "        moduledata = subunitdata[-2]\n",
    "        \n",
    "        # We do this so we can lump in the loading didomain and TE on the first and last modules respectively\n",
    "        modulekeys = list(moduledata.keys())\n",
    "        imodule = 0\n",
    "        while imodule < len(modulekeys):\n",
    "            # Get info\n",
    "            keys = list(moduledata[modulekeys[imodule]].keys())\n",
    "            values = moduledata[modulekeys[imodule]].values()\n",
    "            # Process info according to loading or not\n",
    "            if modules_seen == 0:\n",
    "                loading = True                \n",
    "            else: \n",
    "                loading = False\n",
    "            moduledict = OrderedDict([(k,v) for k,v in zip(keys,values)])\n",
    "            # Determine whether module is terminal or not\n",
    "            if 'Thioesterase' in list(moduledata[modulekeys[imodule]].keys()):\n",
    "                terminal = True\n",
    "            else:\n",
    "                terminal = False\n",
    "            imodule += 1\n",
    "            modules_seen += 1\n",
    "            try:\n",
    "                # This is to make sure we don't add subunits with invalid modules\n",
    "                # The check for errors here is to compare agains the predicted chemcial structure\n",
    "                domains_present = moduledict.keys()\n",
    "                if 'ACP' in domains_present or 'PCP' in domains_present:\n",
    "                    if 'AT' in domains_present or 'ATL' in domains_present:\n",
    "                        # modules.append(pks.Module(moduledict, loading=loading, terminal=terminal))\n",
    "                        module = pks.models.Module(subunit=subunit, loading=loading, terminal=terminal)\n",
    "                        module.save()\n",
    "                        module.buildDomains(moduledict, cyclic=cyclize)\n",
    "            except AssertionError as e:\n",
    "                print(moduledict)\n",
    "                print(type(e).__name__, e.args, subunit + ' ' + subunitdata[1])\n",
    "                raise Exception(type(e).__name__, e.args, subunit + ' ' + subunitdata[1])\n",
    "                break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Process all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0: ('BGC0000971', 'FR687018')\n",
      "['Cinnabaramide']\n",
      "1: ('BGC0000170', 'ACJE01000012')\n",
      "['yanuthone D']\n",
      "2: ('BGC0000958', 'NZ_FR873698')\n",
      "['Antimycin']\n",
      "3: ('BGC0000013', 'BN001304')\n",
      "['alternariol']\n",
      "4: ('BGC0001042', 'FJ809786')\n",
      "['sanglifehrin A']\n",
      "5: ('BGC0000116', 'EU108007')\n",
      "['nystatin-like Pseudonocardia polyene']\n",
      "6: ('BGC0000083', 'GQ274954')\n",
      "['Lactimidomycin',\n",
      " '8,9-dihydro-LTM',\n",
      " '8,9-dihydro-8S-hydroxy-LTM',\n",
      " '8,9-dihydro-9R-hydroxy-LTM']\n",
      "7: ('BGC0000992', 'AY700570')\n",
      "['fusaridione A']\n",
      "8: ('BGC0001172', 'KF585133')\n",
      "['Chlorizidine A']\n",
      "9: ('BGC0001000', 'AY974560')\n",
      "['hectochlorin']\n",
      "10: ('BGC0001160', 'CP000951')\n",
      "['1-nonadecene', '(14Z)-1,14-Nonadecadiene']\n",
      "11: ('BGC0000969', 'AM179409')\n",
      "['Chondramid A']\n",
      "12: ('BGC0001243', 'FR717895')\n",
      "['botryenalol',\n",
      " 'botrydial',\n",
      " 'dihydrobotrydial',\n",
      " 'botryendial',\n",
      " 'beta-O-Methyl-dihydrobotrydialone',\n",
      " 'botcinic acid',\n",
      " 'botcinin A']\n",
      "13: ('BGC0000003', 'AB179766')\n",
      "['AF-toxin']\n",
      "14: ('BGC0001359', 'LN997801')"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 1780, found 19546 (BGC0001243.1).\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 99360, found 191877 (BGC0001359.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "['PM100117', 'PM100118']\n",
      "15: ('BGC0001258', 'AF549411')\n",
      "['1,8-dihydroxynaphthalene']\n",
      "16: ('BGC0000035', 'AY509120')\n",
      "['chalcomycin']\n",
      "17: ('BGC0000121', 'KC145148')\n",
      "['pestheic acid']\n",
      "18: ('BGC0000141', 'AJ871581')\n",
      "['rubradirin']\n",
      "19: ('BGC0001101', 'AF484556')\n",
      "['leinamycin']\n",
      "20: ('BGC0000148', 'AY007564')\n",
      "['spinosad']\n",
      "21: ('BGC0000029', 'FJ872523')\n",
      "['BE-14106']\n",
      "22: ('BGC0001119', 'HF563079.1')\n",
      "['divergolide A', 'divergolide B', 'divergolide C', 'divergolide D']\n",
      "23: ('BGC0000979', 'GQ385959')\n",
      "['cylindrospermopsin']\n",
      "24: ('BGC0000159', 'EF990140')\n",
      "['tautomycin']\n",
      "25: ('BGC0001300', 'KT368179')\n",
      "['anthracimycin']\n",
      "26: ('BGC0001033', 'CP003355')\n",
      "['paenilamicin']\n",
      "27: ('BGC0001236', 'KT282101')\n",
      "['thiotetroamide']\n",
      "28: ('BGC0001064', 'JX477167')\n",
      "['Cylindrocyclophane D', 'Cylindrocyclophane E', 'Cylindrocyclophane F']\n",
      "29: ('BGC0000039', 'AB072893')\n",
      "['compactin']\n",
      "30: ('BGC0000064', 'AY604568')\n",
      "['fusarin']\n",
      "31: ('BGC0000043', 'GQ412749')\n",
      "['curacin A']\n",
      "32: ('BGC0001051', 'GQ981380')\n",
      "['Thuggacin A']\n",
      "33: ('BGC0000056', 'AY267372')\n",
      "['esperamicin']\n",
      "34: ('BGC0001353', 'LN879412')\n",
      "['thiolactomycin']\n",
      "35: ('BGC0000050', 'DQ272520')\n",
      "['E-837']\n",
      "36: ('BGC0001348', 'NZ_JOBF01000033.1')\n",
      "['JBIR-100']\n",
      "37: ('BGC0001301', 'LN871452')\n",
      "['anthracimycin']\n",
      "38: ('BGC0000034', 'AY310323')\n",
      "['candicidin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 13567, found 16305 (BGC0001353.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39: ('BGC0001013', 'DL080157')\n",
      "['meridamycin']\n",
      "40: ('BGC0000980', 'GQ385960')\n",
      "['cylindrospermopsin']\n",
      "41: ('BGC0000020', 'AF453501')\n",
      "['ansamitocin']\n",
      "42: ('BGC0000155', 'FJ943499')\n",
      "['T-toxin']\n",
      "43: ('BGC0001021', 'AL123456')\n",
      "['mycobactin']\n",
      "44: ('BGC0000011', 'AY092402')\n",
      "['aflatoxin', 'sterigmatocystin']"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 5129, found 61691 (BGC0000155.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "45: ('BGC0000054', 'AY623658')\n",
      "['erythromycin']\n",
      "46: ('BGC0000031', 'AJ580915')\n",
      "['Borrelidin']\n",
      "47: ('BGC0001011', 'DQ351275')\n",
      "['meridamycin']\n",
      "48: ('BGC0001161', 'KF550301')\n",
      "['1-heptadecene']\n",
      "49: ('BGC0001218', 'KP737857')\n",
      "['fumosorinone']\n",
      "50: ('BGC0001055', 'AM236324')\n",
      "['yersiniabactin']\n",
      "51: ('BGC0000072', 'KM361622')\n",
      "['gulmirecin A']\n",
      "52: ('BGC0000137', 'CP000850')\n",
      "['rifamycin']\n",
      "53: ('BGC0000995', 'HM047288')\n",
      "['FR901464']\n",
      "54: ('BGC0000017', 'FJ477836')\n",
      "['Anatoxin-a', 'Homoanatoxin-a']\n",
      "55: ('BGC0001195', 'KP339942')\n",
      "['nocardiopsin']\n",
      "56: ('BGC0000084', 'JQ793783')\n",
      "['laidlomycin']\n",
      "57: ('BGC0000972', 'AM229678')\n",
      "['N-myristoyl-D-asparagine',\n",
      " 'cis-7-tetradecenoyl-D-asparagine',\n",
      " '(R)-N1-((S)-5-oxohexan-2-yl)-2-tetradecanamidosuccinamide']\n",
      "58: ('BGC0001390', 'LC125467')\n",
      "['LL-Z1272beta']\n",
      "59: ('BGC0001017', 'AF183408.1')\n",
      "['microcystin LR']\n",
      "60: ('BGC0001262', 'AB872924')\n",
      "['AK-toxin']\n",
      "61: ('BGC0001212', 'KT067736')\n",
      "['nannocystin a']\n",
      "62: ('BGC0000380', 'HM639990')\n",
      "['Leupyrrin']\n",
      "63: ('BGC0000118', 'AB303063')\n",
      "['pactamycin']\n",
      "64: ('BGC0000028', 'GU390405.1')\n",
      "['Bafilomycin B1']\n",
      "65: ('BGC0000136', 'AF040570')\n",
      "['rifamycin']\n",
      "66: ('BGC0000970', 'AM988861')\n",
      "['Chondrochloren']\n",
      "67: ('BGC0000069', 'KF479198')\n",
      "['gephyronic acid']\n",
      "68: ('BGC0001254', 'AB516321')\n",
      "['ACT-Toxin II']\n",
      "69: ('BGC0000158', 'EU035755')\n",
      "['tautomycetin']\n",
      "70: ('BGC0000166', 'U78289')\n",
      "['tylactone']\n",
      "71: ('BGC0001396', 'KU568466')\n",
      "['aldgamycin J', 'aldgamycin K', 'aldgamycin P', 'aldgamycin E']\n",
      "72: ('BGC0001232', 'NZ_JAIY01000017')\n",
      "['leucanicidin']\n",
      "73: ('BGC0000966', 'JF419316')\n",
      "['caerulomycin A']\n",
      "74: ('BGC0001231', 'KF657739')\n",
      "['Microsclerodermin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 40233, found 64266 (BGC0001232.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "75: ('BGC0001144', 'GG698521')\n",
      "['neosartoricin']\n",
      "76: ('BGC0000004', 'AB196490')\n",
      "['aflatoxin']\n",
      "77: ('BGC0000977', 'AB506492')\n",
      "['cyclopiazonic acid']\n",
      "78: ('BGC0000988', 'AF210843')\n",
      "['epothilone']\n",
      "79: ('BGC0000086', 'AB449340')\n",
      "['lasalocid']\n",
      "80: ('BGC0000100', 'AF440781')\n",
      "['monensin']\n",
      "81: ('BGC0000007', 'AY510452')\n",
      "['aflatoxin']\n",
      "82: ('BGC0000142', 'CP000667')\n",
      "['salinilactam']\n",
      "83: ('BGC0000008', 'AY510453')\n",
      "['aflatoxin']\n",
      "84: ('BGC0000146', 'AB514562')\n",
      "['solanapyrone']\n",
      "85: ('BGC0000079', 'FJ545274')\n",
      "['indanomycin']\n",
      "86: ('BGC0000068', 'DQ914285')\n",
      "['geldanamycin']\n",
      "87: ('BGC0000134', 'EU520419')\n",
      "['radicicol']\n",
      "88: ('BGC0000129', 'CM000174')\n",
      "['pyripyropene']\n",
      "89: ('BGC0001287', 'LN831790')\n",
      "['chaxamycin A', 'chaxamycin B', 'chaxamycin C', 'chaxamycin D']\n",
      "90: ('BGC0001022', 'AF319998')\n",
      "['myxalamid']\n",
      "91: ('BGC0000108', 'HQ386234')\n",
      "['natamycin']\n",
      "92: ('BGC0000954', 'AM946600')\n",
      "['Ajudazol']\n",
      "93: ('BGC0001066', 'AM992894')\n",
      "['Kendomycin']\n",
      "94: ('BGC0000058', 'AB469193')\n",
      "['FD-891']\n",
      "95: ('BGC0001355', 'LN879416')\n",
      "['U-68204']\n",
      "96: ('BGC0000430', 'CP007203')\n",
      "['spumigin E']\n",
      "97: ('BGC0001056', 'HE995400')\n",
      "['zeamine', 'zeamine I', 'zeamine II', 'Prezeamine', 'Prezeamine I']\n",
      "98: ('BGC0001305', 'HF679030')\n",
      "['fujikurins']\n",
      "99: ('BGC0000164', 'AB193609')\n",
      "['tetronomycin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 27100, found 30965 (BGC0001355.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100: ('BGC0001203', 'KP830094')\n",
      "['clarexpoxcin']\n",
      "101: ('BGC0000098', 'DQ176595')\n",
      "['monacolin K']\n",
      "102: ('BGC0000095', 'KC608848')\n",
      "['micromonolactam']\n",
      "103: ('BGC0001245', 'KJ434938')\n",
      "['lasiodiplodin']\n",
      "104: ('BGC0001304', 'EQ963475')\n",
      "['aflavarin']\n",
      "105: ('BGC0000102', 'AB089954')\n",
      "['mycinamicin']\n",
      "106: ('BGC0000002', 'CP007155')\n",
      "['aculeximycin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 2254, found 104490 (BGC0000095.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "107: ('BGC0000090', 'EU827593')\n",
      "['Macbecin I', 'Macbecin II']\n",
      "108: ('BGC0000076', 'EU520417')\n",
      "['hypothemycin']\n",
      "109: ('BGC0001394', 'CP004025')\n",
      "['phenalamide']\n",
      "110: ('BGC0000005', 'AF452809')\n",
      "['aflatoxin']\n",
      "111: ('BGC0001054', 'FN667742')\n",
      "['Xenocoumacin 1', 'Xenocoumacin II']\n",
      "Missing subunit: \"XNC1_1704\"\n",
      "['Ketosynthase', 'Acyltransferase', 'Ketoreductase', 'Thiolation (ACP/PCP)']\n",
      "dict_keys(['CBJ89760.1', 'CBJ89766.1', 'CBJ89764.1'])\n",
      "112: ('BGC0001360', 'HG792019')\n",
      "['mycophenolic acid']\n",
      "113: ('BGC0000037', 'BN001301')\n",
      "['cichorine']\n",
      "114: ('BGC0001169', 'KF874660.1')\n",
      "['piericidin A1']\n",
      "115: ('BGC0000105', 'AF521085')\n",
      "['nanchangmycin']\n",
      "116: ('BGC0000042', 'AB818354')\n",
      "['cremimycin']\n",
      "117: ('BGC0000171', 'FR878059')\n",
      "['9-methylstreptimidone']\n",
      "118: ('BGC0001031', 'JX315603')\n",
      "['oocydin A']\n",
      "119: ('BGC0001027', 'AP006618')\n",
      "['nocobactin NA']\n",
      "120: ('BGC0001270', 'Y15279')\n",
      "['gibberellin A3',\n",
      " 'gibberellin A1',\n",
      " 'gibberellin A7',\n",
      " 'gibberellin A4',\n",
      " 'gibberellin A9']\n",
      "121: ('BGC0001373', 'LC079035')\n",
      "['sordarin']\n",
      "122: ('BGC0001397', 'CP013739')\n",
      "['lidamycin']\n",
      "123: ('BGC0000023', 'AM850130')\n",
      "['Aurafuron A']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 2900, found 23655 (BGC0001270.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "124: ('BGC0001246', 'KJ434939')\n",
      "['trans-resorcylide']\n",
      "125: ('BGC0000077', 'EU520418')\n",
      "['hypothemycin']\n",
      "126: ('BGC0001281', 'NC_026500')\n",
      "['ustilagic acid']\n",
      "127: ('BGC0000135', 'AB568601')\n",
      "['reveromycin']\n",
      "128: ('BGC0001032', 'JX315604')\n",
      "['oocydin A']\n",
      "129: ('BGC0000177', 'GQ274953')\n",
      "['iso-migrastatin',\n",
      " 'migrastatin',\n",
      " 'Dorrigocin A',\n",
      " 'Dorrigocin B',\n",
      " '13-epi-Dorrigocin A']\n",
      "130: ('BGC0001005', 'KF866134')\n",
      "['locillomycin']\n",
      "131: ('BGC0001253', 'AB444613')\n",
      "['ACT-Toxin I']\n",
      "132: ('BGC0001383', 'KF815729')\n",
      "['macrolactin 1a', 'macrolactin 1b', 'macrolactin 1c', 'macrolactin 3a']\n",
      "133: ('BGC0001125', 'KM078884')\n",
      "['puwainaphycin A', 'puwainaphycin B', 'puwainaphycin C', 'puwainaphycin D']\n",
      "134: ('BGC0000052', 'AY899214')\n",
      "['ECO-02301']\n",
      "135: ('BGC0000162', 'EU443633')\n",
      "['tetrocarcin A']\n",
      "136: ('BGC0001070', 'AM746336')\n",
      "['kirromycin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 2773, found 49372 (BGC0001253.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "137: ('BGC0000981', 'GQ385961')\n",
      "['cylindrospermopsin']\n",
      "138: ('BGC0001276', 'D85860')\n",
      "['6-methylsalicyclic acid']\n",
      "139: ('BGC0000117', 'AB070940')\n",
      "['oligomycin']\n",
      "140: ('BGC0000099', 'HM070047')\n",
      "['monascorubrin']\n",
      "141: ('BGC0001045', 'JQ045344')\n",
      "['Spiruchostatin A']\n",
      "142: ('BGC0001358', 'LSBH01000002')\n",
      "['leucinostatin A', 'leucinostatin B']\n",
      "143: ('BGC0000143', 'DQ630728')\n",
      "['salinomycin']\n",
      "144: ('BGC0000082', 'EU301739')\n",
      "['kijanimicin']\n",
      "145: ('BGC0001403', 'CM000172')\n",
      "['trypacidin']\n",
      "146: ('BGC0001349', 'KP742963')\n",
      "['heronamide A',\n",
      " 'heronamide B',\n",
      " 'heronamide C',\n",
      " 'heronamide D',\n",
      " 'heronamide E',\n",
      " 'heronamide F']\n",
      "147: ('BGC0000132', 'EF140903')\n",
      "['pyrrolomycin']\n",
      "148: ('BGC0001381', 'KP161205')\n",
      "['brasilinolide A', 'brasilinolide B', 'brasilinolide C']\n",
      "149: ('BGC0001159', 'JX157625')\n",
      "['marinopyrrole A',\n",
      " 'marynopyrrole E',\n",
      " 'marynopyrrole D',\n",
      " 'marynopyrrole C',\n",
      " 'marinopyrrole B']\n",
      "150: ('BGC0000073', 'AB241068')\n",
      "['halstoctacosanolide']\n",
      "151: ('BGC0001266', 'GU930713')\n",
      "['grayanic acid']\n",
      "152: ('BGC0000093', 'FJ952082')\n",
      "['meilingmycin']\n",
      "153: ('BGC0001023', 'AJ698723')\n",
      "['Myxochromide S1']\n",
      "154: ('BGC0000024', 'AJ575648')\n",
      "['aureothin']\n",
      "155: ('BGC0000124', 'HQ840721')\n",
      "['piericidin A1']\n",
      "156: ('BGC0001047', 'AJ548826')\n",
      "['syringolin A']\n",
      "157: ('BGC0001003', 'DQ176871.1')\n",
      "['alpha-lipomycin']\n",
      "158: ('BGC0001002', 'AB920328')\n",
      "['jawsamycin']\n",
      "159: ('BGC0000094', 'AF079138')\n",
      "['methymycin']\n",
      "160: ('BGC0001344', 'GU002154')\n",
      "['tubulysin A',\n",
      " 'tubulysin B',\n",
      " 'tubulysin C',\n",
      " 'tubulysin D',\n",
      " 'tubulysin E',\n",
      " 'tubulysin F',\n",
      " 'tubulysin G',\n",
      " 'tubulysin H',\n",
      " 'tubulysin I']\n",
      "161: ('BGC0001068', 'FW556769')\n",
      "['pyripyropene A']\n",
      "162: ('BGC0000006', 'AY510451')\n",
      "['aflatoxin']\n",
      "163: ('BGC0000974', 'FN547928')\n",
      "['crocacin']\n",
      "164: ('BGC0001382', 'KT362050')\n",
      "['svaricin']\n",
      "165: ('BGC0000097', 'FJ872525')\n",
      "['ML-449']\n",
      "166: ('BGC0001059', 'FJ430564')\n",
      "['zwittermycin A']\n",
      "167: ('BGC0001235', 'KU245058')\n",
      "['antalid']\n",
      "168: ('BGC0001350', 'KT727016')\n",
      "['phormidolide']\n",
      "169: ('BGC0000986', 'BN001209')\n",
      "['Dkxanthene']\n",
      "170: ('BGC0001015', 'AJ441056')\n",
      "['microcystin']\n",
      "171: ('BGC0000133', 'JF970188')\n",
      "['quartromicin']\n",
      "172: ('BGC0000152', 'U34740')\n",
      "['sterigmatocystin']\n",
      "173: ('BGC0000088', 'AF141924')\n",
      "['lovastatin']\n",
      "174: ('BGC0001164', 'CP002198')\n",
      "['1-heptadecene']\n",
      "175: ('BGC0001062', 'FJ915123')\n",
      "['TLN-05220 ']\n",
      "176: ('BGC0000990', 'GU063811')\n",
      "['epothilone']\n",
      "177: ('BGC0000111', 'AB097904')\n",
      "['neocarzilin']\n",
      "178: ('BGC0000009', 'AY510454')\n",
      "['aflatoxin']\n",
      "179: ('BGC0001143', 'ACJE01000001')\n",
      "['azanigerone']\n",
      "180: ('BGC0000075', 'JX504844')\n",
      "['hygrocin A', 'hygrocin B']\n",
      "181: ('BGC0001363', 'AL646053')\n",
      "['ralsolamycin']\n",
      "182: ('BGC0000103', 'BX649209')\n",
      "['mycolactone']\n",
      "183: ('BGC0000049', 'DQ272521')\n",
      "['E-492', 'E-975']\n",
      "184: ('BGC0001342', 'GQ981381')\n",
      "['Cmc-thuggacin A', 'Cmc-thuggacin B']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 35351, found 72330 (BGC0000049.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "185: ('BGC0000955', 'FR831800')\n",
      "['Althiomycin']\n",
      "186: ('BGC0000110', 'AM778535')\n",
      "['Neoaureothin', 'orinocin', 'SNF4435C', 'SNF4435D']\n",
      "187: ('BGC0000040', 'DQ149987')\n",
      "['concanamycin A']\n",
      "188: ('BGC0001065', 'JN671974')\n",
      "['herboxidiene']\n",
      "189: ('BGC0001053', 'AJ620477')\n",
      "['tubulysin']\n",
      "190: ('BGC0000173', 'JX173632')\n",
      "['bongkrekic acid', 'isobongkrekic acid']\n",
      "191: ('BGC0000125', 'AJ278573')\n",
      "['pimaricin']\n",
      "192: ('BGC0001273', 'AY540947')\n",
      "['asperlactone']\n",
      "193: ('BGC0000978', 'EU140798')\n",
      "['cylindrospermopsin']\n",
      "194: ('BGC0000963', 'AF210249')\n",
      "['bleomycin']\n",
      "195: ('BGC0000973', 'HE575208')\n",
      "['collismycin A']\n",
      "196: ('BGC0000062', 'AF155773')\n",
      "['fumonisin']\n",
      "197: ('BGC0001265', 'AB176546')\n",
      "['melanin']\n",
      "198: ('BGC0001034', 'HE616533')\n",
      "['Pellasoren']\n",
      "199: ('BGC0000160', 'CH476602')\n",
      "['terreic acid']\n",
      "200: ('BGC0000140', 'GQ332353')\n",
      "['RK-682']\n",
      "201: ('BGC0001216', 'KP719128')\n",
      "['splenocin']\n",
      "202: ('BGC0000127', 'AF081920')\n",
      "['pyoluteorin']\n",
      "203: ('BGC0000001', 'JF752342')\n",
      "['Abyssomicin C', 'Atrop-abyssomicin C']\n",
      "204: ('BGC0001237', 'KT282100')\n",
      "['thiolactomycin']\n",
      "205: ('BGC0000126', 'AB435553')\n",
      "['pladienolide']\n",
      "206: ('BGC0001299', 'KT368180')\n",
      "['chlorotonil']\n",
      "207: ('BGC0001165', 'AY652953')\n",
      "['Curacin A']\n",
      "Missing subunit: \"AAT70096.1\"\n",
      "['AT', 'T']\n",
      "['KS', 'AT', 'T', 'T', 'T']\n",
      "dict_keys(['AAT70103.1', 'AAT70108.1', 'AAT70102.1', 'AAT70106.1', 'AAT70105.1', 'AAT70107.1', 'AAT70104.1'])\n",
      "208: ('BGC0001043', 'AP009493')\n",
      "['SGR PTMs']\n",
      "209: ('BGC0000998', 'JX545234')\n",
      "['guadinomine']\n",
      "210: ('BGC0001040', 'X86780')\n",
      "['rapamycin']\n",
      "211: ('BGC0001309', 'KT363732')\n",
      "['usnic acid']\n",
      "212: ('BGC0001303', 'KT209587')\n",
      "['lobosamide']\n",
      "213: ('BGC0001255', 'KC439347')\n",
      "['equisetin']\n",
      "214: ('BGC0001004', 'KC013978')\n",
      "['lobophorin B']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 7950, found 9961 (BGC0001309.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "215: ('BGC0000128', 'AY394844')\n",
      "['pyoluteorin']\n",
      "216: ('BGC0001378', 'KT716443')\n",
      "['tiancimycin']\n",
      "217: ('BGC0000059', 'BA000030')\n",
      "['filipin']\n",
      "218: ('BGC0000984', 'CP000884')\n",
      "['delftibactin A', 'delftibactin B']\n",
      "219: ('BGC0001264', 'LC011911')\n",
      "['betaenone A', 'betaenone B', 'betaenone C']\n",
      "220: ('BGC0001010', 'AJ557546')\n",
      "['Melithiazol A']\n",
      "221: ('BGC0001252', 'AB725683')\n",
      "['ACR-toxin']\n",
      "222: ('BGC0001202', 'KP830093')\n",
      "['landepoxcin']\n",
      "223: ('BGC0001163', 'GL890825')\n",
      "['1-heptadecene']\n",
      "224: ('BGC0000014', 'DQ897667')\n",
      "['ambruticin']\n",
      "225: ('BGC0000012', 'AB120221')\n",
      "['alternapyrone']\n",
      "226: ('BGC0000018', 'EU220288')\n",
      "['angolamycin']\n",
      "227: ('BGC0000163', 'FJ462704')\n",
      "['tetronasin']\n",
      "228: ('BGC0001401', 'JF739169')\n",
      "['melleolide F', \"6'-Chloromelleolide F\", \"6'-Bromomelleolide F\"]\n",
      "229: ('BGC0001257', 'AF151533')\n",
      "['1,3,6,8-tetrahydroxynaphthalene']\n",
      "230: ('BGC0001400', 'CH476608')\n",
      "['citreoviridin']\n",
      "231: ('BGC0000055', 'AM420293')\n",
      "['Erythromycin A', 'Erythromycin B', 'Erythromycin C', 'Erythromycin D']\n",
      "232: ('BGC0000167', 'AB086653')\n",
      "['vicenistatin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 2055, found 10531 (BGC0001401.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "233: ('BGC0001001', 'AY522504')\n",
      "['Jamaicamide', 'Jamaicamide b', 'jamaicamide c']\n",
      "234: ('BGC0000122', 'KF739396')\n",
      "['phenylnannolone A', 'phenylnannolone B', 'phenylnannolone C']\n",
      "235: ('BGC0001183', 'JX306680')\n",
      "['lobophorin']\n",
      "236: ('BGC0000106', 'GQ452266')\n",
      "['naphthomycin']\n",
      "237: ('BGC0001367', 'CP001804')\n",
      "['haliamide']\n",
      "238: ('BGC0001296', 'LC061217')\n",
      "['streptazone E']\n",
      "239: ('BGC0001405', 'CH408033')\n",
      "['chaetoviridin', 'chaetomugilin']\n",
      "240: ('BGC0001012', 'DQ885223')\n",
      "['meridamycin']\n",
      "241: ('BGC0001293', 'KT327068')\n",
      "['cyclizidine']\n",
      "242: ('BGC0000131', 'EF140902')\n",
      "['pyrrolomycin']\n",
      "243: ('BGC0000047', 'AY118081')\n",
      "['dihydrochalcomycin']\n",
      "244: ('BGC0000063', 'EU449979')\n",
      "['fumonisin']\n",
      "245: ('BGC0000993', 'EF210776')\n",
      "['FK228']\n",
      "246: ('BGC0000952', 'FR681999.1')\n",
      "['pristinamycin']\n",
      "247: ('BGC0000078', 'AB767280')\n",
      "['incednine']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 66339, found 235448 (BGC0000952.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "248: ('BGC0001044', 'AY553235')\n",
      "['sirodesmin']\n",
      "249: ('BGC0000030', 'HF679027')\n",
      "['bikaverin']\n",
      "250: ('BGC0001298', 'KF683117')\n",
      "['4-Z-annimycin']\n",
      "251: ('BGC0000147', 'U24241')\n",
      "['Soraphen']\n",
      "252: ('BGC0000061', 'AY310323')\n",
      "['FR-008']\n",
      "253: ('BGC0001272', 'EU086466')\n",
      "['elsinochrome B', 'elsinochrome C']\n",
      "254: ('BGC0000022', 'AACD01000015')\n",
      "['asperfuranone']\n",
      "255: ('BGC0001102', 'CP000085')\n",
      "['Malleilactone']\n",
      "256: ('BGC0001268', 'JX308619')\n",
      "['fusarin']\n",
      "257: ('BGC0000982', 'AY834753')\n",
      "['cystothiazole A']\n",
      "258: ('BGC0000021', 'JF819834')\n",
      "['apoptolidin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 8031, found 30617 (BGC0001272.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "259: ('BGC0001230', 'BK009377')\n",
      "['Salinamide A',\n",
      " 'Salinamide B',\n",
      " 'Salinamide C',\n",
      " 'Salinamide D',\n",
      " 'Salinamide E',\n",
      " 'Salinamide F',\n",
      " 'Desmethylsalinamide C',\n",
      " 'Desmethylsalinamide E']\n",
      "260: ('BGC0000115', 'AF263912')\n",
      "['nystatin A1']\n",
      "261: ('BGC0001024', 'AF188287')\n",
      "['Myxothiazol']\n",
      "262: ('BGC0001269', 'AM886292')\n",
      "['gibberellin A3',\n",
      " 'gibberellin A1',\n",
      " 'gibberellin A7',\n",
      " 'gibberellin A4',\n",
      " 'gibberellin A9']\n",
      "263: ('BGC0001038', 'JX424761')\n",
      "['pyralomicin 1a']\n",
      "264: ('BGC0000383', 'BX470251')\n",
      "['Luminmycin']\n",
      "265: ('BGC0001247', 'AY649543')\n",
      "['cercosporin']\n",
      "266: ('BGC0001057', 'DQ019316')\n",
      "['zearalenone']\n",
      "267: ('BGC0001058', 'EU670723')\n",
      "['zorbamycin']\n",
      "Missing subunit: \"ACG60781.1\"\n",
      "['KS', 'AT', 'KR', 'T']\n",
      "dict_keys(['ACG60777.1', 'ACG60770.1', 'ACG60783.1'])\n",
      "268: ('BGC0000089', 'AF141925')\n",
      "['lovastatin']\n",
      "269: ('BGC0001060', 'EF552206')\n",
      "['dynemicin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 4996, found 17156 (BGC0001269.1).\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 6591, found 36089 (BGC0001247.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "270: ('BGC0000997', 'AM412319')\n",
      "['glidobactin']\n",
      "271: ('BGC0000025', 'AB032523')\n",
      "['avermectin']\n",
      "272: ('BGC0000153', 'AJ421825')\n",
      "['Stigmatellin A']\n",
      "273: ('BGC0000989', 'EU414841')\n",
      "['epothilone']\n",
      "274: ('BGC0000048', 'DQ149246')\n",
      "['dothistromin']\n",
      "275: ('BGC0000994', 'AF235504')\n",
      "['FK520']\n",
      "276: ('BGC0000060', 'HQ434551')\n",
      "['fostriecin']\n",
      "277: ('BGC0001099', 'GU479979')\n",
      "['Kalimantacin A']\n",
      "278: ('BGC0001162', 'KF550302')\n",
      "['1-heptadecene']\n",
      "279: ('BGC0000346', 'KF647219')\n",
      "['epoxomicin']\n",
      "280: ('BGC0000091', 'KF711829')\n",
      "['marineosin']\n",
      "281: ('BGC0001052', 'GU385216')\n",
      "['tirandamycin']\n",
      "282: ('BGC0001020', 'HQ257512')\n",
      "['muraymycin']\n",
      "283: ('BGC0000186', 'CP000085')\n",
      "['thailandamide A', 'thailandamide B', 'thailandamide lactone']\n",
      "284: ('BGC0001018', 'AB481215')\n",
      "['micropeptin K139']\n",
      "285: ('BGC0001352', 'LN879414')\n",
      "['Tu 3010']\n",
      "286: ('BGC0001120', 'CP000086')\n",
      "['burkholderic acid']\n",
      "Missing subunit: \"ABC36203.1\"\n",
      "dict_keys([])\n",
      "287: ('BGC0001244', 'KM365454')\n",
      "['(-)-Mellein']\n",
      "288: ('BGC0001030', 'AY557343')\n",
      "['ochratoxin A']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 46601, found 50467 (BGC0001352.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "289: ('BGC0000138', 'AY442225')\n",
      "['rimocidin']\n",
      "290: ('BGC0000165', 'HQ011923.1')\n",
      "['tiacumicin B']\n",
      "291: ('BGC0000123', 'AY354515')\n",
      "['phoslactomycin B']\n",
      "292: ('BGC0001063', 'AL645882')\n",
      "['undecylprodigiosin']\n",
      "293: ('BGC0001009', 'AJ557546')\n",
      "['melithiazol']\n",
      "294: ('BGC0001048', 'EF032505')\n",
      "['tallysomycin A']\n",
      "295: ('BGC0000149', 'AM407731')\n",
      "['Spirangien A1']\n",
      "296: ('BGC0000961', 'CP000085')\n",
      "['bactobolin']\n",
      "297: ('BGC0000096', 'BD420675')\n",
      "['midecamycin']\n",
      "298: ('BGC0000957', 'KF813023')\n",
      "['ansatrienin (mycotrienin)']\n",
      "299: ('BGC0001036', 'KF386858')\n",
      "['polyoxypeptin']\n",
      "300: ('BGC0001357', 'KT826756')\n",
      "['carbamidocyclophane A',\n",
      " 'carbamidocyclophane B',\n",
      " 'carbamidocyclophane C',\n",
      " 'carbamidocyclophane D',\n",
      " 'carbamidocyclophane E',\n",
      " 'carbamidocyclophane F',\n",
      " 'carbamidocyclophane H',\n",
      " 'carbamidocyclophane M',\n",
      " 'carbamidocyclophane N',\n",
      " 'carbamidocyclophane O',\n",
      " 'carbamidocyclophane P',\n",
      " 'carbamidocyclophane Q',\n",
      " 'carbamidocyclophane R',\n",
      " 'carbamidocyclophane S',\n",
      " 'carbamidocyclophane T',\n",
      " 'carbamidocyclophane U']\n",
      "301: ('BGC0000053', 'GP697151')\n",
      "['elaiophylin']\n",
      "302: ('BGC0000036', 'DQ116941')\n",
      "['chlorothricin', 'deschlorothricin']\n",
      "303: ('BGC0000092', 'AF263245')\n",
      "['megalomicin A', 'megalomicin B', 'megalomicin C1', 'megalomicin C2']\n",
      "304: ('BGC0000976', 'HQ696500')\n",
      "['Curacin A']\n",
      "305: ('BGC0000151', 'AM238664')\n",
      "['stambomycin A', 'stambomycin B', 'stambomycin C', 'stambomycin D']\n",
      "306: ('BGC0000130', 'EF140901')\n",
      "['pyrrolomycin']\n",
      "307: ('BGC0001354', 'LN879418')\n",
      "['U-68204']\n",
      "308: ('BGC0000015', 'AF357202')\n",
      "['amphotericin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 26211, found 29917 (BGC0001354.1).\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 113193, found 135074 (BGC0000015.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "309: ('BGC0000032', 'HM452329')\n",
      "['calcimycin']\n",
      "310: ('BGC0000085', 'AB088224')\n",
      "['lankamycin']\n",
      "311: ('BGC0000109', 'AB363939')\n",
      "['nemadectin']\n",
      "312: ('BGC0001259', 'EU086466')\n",
      "['elsinochrome A']\n",
      "313: ('BGC0000114', 'DQ354110')\n",
      "['nigericin']\n",
      "314: ('BGC0001014', 'AL646052')\n",
      "['micacocidin']\n",
      "315: ('BGC0001046', 'FN433113')\n",
      "['Streptolydigin']\n",
      "316: ('BGC0000010', 'AY510455')\n",
      "['aflatoxin']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 8031, found 30675 (BGC0001259.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "317: ('BGC0000038', 'AL645882')\n",
      "['coelimycin P1']\n",
      "318: ('BGC0000033', 'AF497482')\n",
      "['calicheamicin']\n",
      "319: ('BGC0000080', 'DQ897668')\n",
      "['jerangolid A', 'jerangolid D']\n",
      "320: ('BGC0000019', 'EU232693')\n",
      "['angolamycin']\n",
      "321: ('BGC0000144', 'JN033543')\n",
      "['salinomycin']\n",
      "322: ('BGC0001288', 'LC021382')\n",
      "['maklamicin']\n",
      "323: ('BGC0001019', 'KF657738')\n",
      "['Microsclerodermin M']\n",
      "324: ('BGC0000345', 'KF647220')\n",
      "['eponemycin']\n",
      "325: ('BGC0000066', 'AY179507')\n",
      "['geldanamycin']\n",
      "326: ('BGC0001275', 'AY941322')\n",
      "['6-methylsalicyclic acid']\n",
      "327: ('BGC0000045', 'JX971534')\n",
      "['dehydrocurvularin']\n",
      "328: ('BGC0000087', 'FM173265')\n",
      "['lasalocid']\n",
      "329: ('BGC0000962', 'AF516145')\n",
      "['barbamide']\n",
      "330: ('BGC0001271', 'BN001308')\n",
      "['emericellin']\n",
      "331: ('BGC0000113', 'AF016585')\n",
      "['niddamycin']\n",
      "332: ('BGC0000145', 'EF397502')\n",
      "['salinosporamide A']\n",
      "333: ('BGC0001071', 'GQ979609')\n",
      "['nosperin']\n",
      "334: ('BGC0000991', 'AF217189')\n",
      "['epothilone']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.4/dist-packages/Bio/GenBank/__init__.py:1218: BiopythonParserWarning: Expected sequence length 29911, found 34750 (BGC0001271.1).\n",
      "  BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "335: ('BGC0001072', 'AF324838.2')\n",
      "['Simocyclinone D8']\n",
      "336: ('BGC0000067', 'DQ249341')\n",
      "['geldanamycin']\n",
      "337: ('BGC0001039', 'HM436809')\n",
      "['pyridomycin']\n",
      "338: ('BGC0000157', 'DQ983361')\n",
      "['tautomycetin']\n",
      "339: ('BGC0001385', 'KU928136')\n",
      "['byssochlamic acid', 'agnestadrides']\n",
      "340: ('BGC0000051', 'KC894072')\n",
      "['ebelactone']\n",
      "341: ('BGC0000960', 'EU240558')\n",
      "['azinomycin B']\n",
      "342: ('BGC0000107', 'BN001302')\n",
      "['naphthopyrone']\n",
      "343: ('BGC0000041', 'AF098795')\n",
      "['coronafacic acid']\n",
      "344: ('BGC0000999', 'EF028635')\n",
      "['heat-stable antifungal factor']\n",
      "345: ('BGC0001404', 'AM920436')\n",
      "['sorbicillin']\n",
      "346: ('BGC0001377', 'KT762610')\n",
      "['uncialamycin']\n",
      "347: ('BGC0000393', 'CP000113')\n",
      "['Myxoprincomide-c506']\n",
      "348: ('BGC0001026', 'GQ176852')\n",
      "['NG-391']\n",
      "349: ('BGC0001280', 'LC011911')\n",
      "['betaenone C', 'betaenone A']\n",
      "350: ('BGC0001242', 'HE613440')\n",
      "['fusarubin']\n",
      "351: ('BGC0000074', 'AY947889')\n",
      "['herbimycin']\n"
     ]
    }
   ],
   "source": [
    "# This cleans up the dictionary\n",
    "[cluster.delete() for cluster in pks.models.Cluster.objects.all()]\n",
    "\n",
    "# Iterate over list of type I modular PKSs\n",
    "for i in range(len(t1pks)):\n",
    "    print('%d: %s' %(i, t1pks[i]))\n",
    "    entry = t1pks[i]\n",
    "\n",
    "     # This prints the accession number and product compound of the cluster\n",
    "    with open(os.path.join(file_path, entry[0] + '.json')) as json_file:\n",
    "        mibig_data = json.load(json_file)\n",
    "        pprint([compound['compound'] for compound in mibig_data['general_params']['compounds']])\n",
    "\n",
    "    # Read in cluster data\n",
    "    record = SeqIO.read(os.path.join(antismash_file_path, entry[0] + '.embl'), \"embl\")\n",
    "    try:\n",
    "        cluster_ref = pks.models.Cluster(\n",
    "            genbankAccession=record.annotations['comment'].split()[-1].strip().strip('.'), \\\n",
    "            mibigAccession=record.id, \\\n",
    "            description=record.description.replace(' biosynthetic gene cluster', ''), \\\n",
    "            sequence=record.seq)\n",
    "        cluster_ref.save()\n",
    "        # Processes subunits and modules belonging to cluster\n",
    "        process_cluster(record, cluster_ref, os.path.join(file_path, entry[0] + '.json'))\n",
    "    except Exception:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Manually reorder clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<QuerySet [<Subunit: nbrI pks subunit>, <Subunit: nbrJ pks subunit>, <Subunit: nbrK pks subunit>, <Subunit: nbrL pks subunit>, <Subunit: nbrG pks subunit>, <Subunit: nbrH pks subunit>]>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# BE-14106\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0000029.1')\n",
    "cluster.reorderSubunits(['becB', 'becD', 'becE', 'becF', 'becG'])\n",
    "pks.models.setCyclization(cluster, False)\n",
    "# Cremimycin\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0000042.1')\n",
    "cluster.reorderSubunits(['cmiP7', 'cmiP8', 'cmiP1', 'cmiP4', 'cmiP5', 'cmiP6'])\n",
    "# Meilingmycin\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0000093.1')\n",
    "cluster.deleteSubunit('pks1')\n",
    "cluster.deleteSubunit('pks2')\n",
    "# ML-449\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0000097.1')\n",
    "cluster.reorderSubunits(['mlaB', 'mlaD', 'mlaE', 'mlaF', 'mlaG'])\n",
    "pks.models.setCyclization(cluster, False)\n",
    "# Nystatin\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0000115.1')\n",
    "cluster.reorderSubunits(['nysA', 'nysB', 'nysC', 'nysI', 'nysJ', 'nysK'])\n",
    "# Tiacumicin\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0000165.1')\n",
    "cluster.deleteSubunit('tiaB')\n",
    "# Simocyclinone\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0001072.1')\n",
    "cluster.reorderSubunits(['simC1A', 'simC1B', 'simC1C'])\n",
    "# Brasilinolide\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession='BGC0001381.1')\n",
    "cluster.reorderSubunits(['nbrI', 'nbrJ', 'nbrK', 'nbrL', 'nbrG', 'nbrH'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Testing cluster chemical operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[<Subunit: becB pks subunit>,\n",
      "  [[<Module: pks module 12539>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mal, loading True>, <KR: type U>, <ACP: domain>]>],\n",
      "   [<Module: pks module 12540>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mmal, loading False>, <KR: type B1>, <ACP: domain>]>]]],\n",
      " [<Subunit: becD pks subunit>,\n",
      "  [[<Module: pks module 12537>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mal, loading False>, <KR: type C2>, <ACP: domain>]>],\n",
      "   [<Module: pks module 12538>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mal, loading False>, <KR: type U>, <ACP: domain>]>]]],\n",
      " [<Subunit: becE pks subunit>,\n",
      "  [[<Module: pks module 12533>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mal, loading False>, <KR: type B1>, <ACP: domain>]>]]],\n",
      " [<Subunit: becF pks subunit>,\n",
      "  [[<Module: pks module 12534>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mmal, loading False>, <KR: type B1>, <ACP: domain>]>],\n",
      "   [<Module: pks module 12535>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mal, loading False>, <KR: type B1>, <ACP: domain>]>]]],\n",
      " [<Subunit: becG pks subunit>,\n",
      "  [[<Module: pks module 12536>,\n",
      "    <QuerySet [<KS: domain>, <AT: substrate mal, loading False>, <KR: type U>, <ACP: domain>, <TE: cyclic False>]>]]]]\n",
      "[<rdkit.Chem.rdchem.Mol object at 0x7fe5b8154ba8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b74ee3c8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b73fb668>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b73fb358>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b73fb2e8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b73fb128>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b73fb5f8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b74ee2e8>]\n"
     ]
    }
   ],
   "source": [
    "acc = 'BGC0000029.1'\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession=acc)\n",
    "pprint(cluster.architecture())\n",
    "pprint(cluster.computeProduct())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Validate and then delete clusters lacking valid products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BGC0001011.1: Meridamycin\n",
      "BGC0001212.1: Nannocystin\n",
      "BGC0000954.1: Ajudazol\n",
      "BGC0000164.1: Tetronomycin\n",
      "BGC0001070.1: Kirromycin\n",
      "BGC0001344.1: Tubulysin\n",
      "BGC0000974.1: Crocacin\n",
      "BGC0001059.1: Zwittermycin A\n",
      "BGC0001034.1: Pellasoren\n",
      "BGC0001010.1: Melithiazol\n",
      "BGC0000163.1: Tetronasin\n",
      "BGC0001293.1: Cyclizidine\n",
      "BGC0000982.1: Cystothiazole A\n",
      "BGC0000021.1: Apoptolidin\n",
      "BGC0001024.1: Myxothiazol\n",
      "BGC0000153.1: Stigmatellin\n"
     ]
    }
   ],
   "source": [
    "# Delete clusters with no computable product\n",
    "for cluster in pks.models.Cluster.objects.all():\n",
    "    try:\n",
    "        cluster.computeProduct()\n",
    "    except:\n",
    "        acc = cluster.mibigAccession\n",
    "        print('%s: %s' %(acc, cluster.description))\n",
    "        cluster.delete()\n",
    "#        cluster = pks.models.Cluster.objects.get(mibigAccession=acc)\n",
    "#        cluster.computeProduct()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Validate and then delete clusters with less than three modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Delete clusters with no modules\n",
    "for cluster in pks.models.Cluster.objects.all():\n",
    "    for subunit in cluster.subunits():\n",
    "        if len(subunit.modules()) == 0:\n",
    "            subunit.delete()\n",
    "    if len(cluster.subunits()) == 0:\n",
    "        cluster.delete()\n",
    "        \n",
    "# Delete clusters with less than three modules\n",
    "for cluster in pks.models.Cluster.objects.all():\n",
    "    nmodules = 0\n",
    "    for subunit in cluster.subunits():\n",
    "        nmodules += len(subunit.modules())\n",
    "    if nmodules < 3:\n",
    "        cluster.delete()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Compare Django object clusters with IPython Notebook object clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "notebook_avail = ['LC021382.1', 'AL645882.2', 'JN671974.1', 'AB089954.2', 'AY834753.1', 'AF453501.1', 'FN547928.1', 'HQ386234.1', 'FJ872523.1', 'HQ434551.1', 'HM452329.1', 'AF217189.1', 'JX504844.1', 'AM179409.1', 'AF016585.1', 'JF819834.1', 'KU245058.1', 'AJ557546.1', 'LC061217.1', 'AJ441056.1', 'AY354515.1', 'AB568601.1', 'KT209587.1', 'AF079138.1', 'AM420293.1', 'FJ545274.1', 'AY007564.1', 'HM639990.1', 'CP007155.1', 'AB449340.1', 'KU568466.1', 'AB767280.1', 'AF521085.1', 'EU035755.1', 'EF990140.1', 'GQ981381.1', 'EU108007.1', 'JX545234.1', 'KF585133.1', 'KT327068.1', 'X86780.1', 'JQ793783.1', 'AY623658.2', 'AJ871581.1', 'AB818354.1', 'AF319998.1', 'BA000030.3', 'EU220288.1', 'DQ885223.1', 'AF440781.1', 'AF188287.1', 'DQ351275.1', 'AB086653.1', 'AF263912.1', 'KC894072.1', 'KT067736.1', 'AJ421825.1', 'AB435553.1', 'DQ176871.1', 'U78289.1', 'AB241068.1', 'FJ809786.1', 'AM850130.1', 'KM361622.1', 'EU414841.1', 'BX649209.1', 'FJ872525.1', 'AJ278573.1', 'AY179507.1', 'AY509120.1', 'AF324838.2', 'AB088224.2', 'GQ385961.1', 'GQ452266.1', 'FN433113.1', 'AY118081.2', 'KC013978.1', 'BN001209.1', 'DQ897668.1', 'FJ952082.1', 'AF040570.3', 'DQ897667.1', 'AB469193.1', 'GU063811.1', 'AF098795.1', 'AY310323.2', 'AY310323.2', 'AB363939.1', 'LN997801.1', 'JN033543.1', 'FM173265.1', 'KF479198.1', 'EU827593.1', 'AM407731.1', 'AJ557546.1', 'JF752342.1', 'AF357202.1', 'AJ580915.1', 'GQ981380.1', 'HQ011923.1', 'AF263245.1', 'AM778535.1', 'AF235504.1', 'LN831790.1', 'AY947889.1', 'EU140798.1', 'KP161205.1', 'U24241.2', 'AF210843.1']\n",
    "notebook_avail = set(notebook_avail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "django_avail = []\n",
    "for cluster in pks.models.Cluster.objects.all():\n",
    "    django_avail.append(cluster.genbankAccession)\n",
    "django_avail = set(django_avail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'DQ351275.1', 'KT327068.1', 'AY834753.1', 'AJ421825.1', 'AJ557546.1', 'KT067736.1', 'JF819834.1', 'AF188287.1', 'FN547928.1', 'AL645882.2'}\n",
      "{'KF739396.1', 'AM988861.1', 'HF563079.1'}\n"
     ]
    }
   ],
   "source": [
    "only_notebook = notebook_avail.difference(django_avail)\n",
    "only_django = django_avail.difference(notebook_avail)\n",
    "print(only_notebook)\n",
    "print(only_django)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Compare Django object clusters with available final structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n"
     ]
    }
   ],
   "source": [
    "django_avail = []\n",
    "for cluster in pks.models.Cluster.objects.all():\n",
    "    django_avail.append(cluster.mibigAccession.split('.')[0])\n",
    "django_avail = set(django_avail)\n",
    "struct_avail = set(pickle.load(open('./mibig_rdkitmol_finalstructs.p', 'rb')).keys())\n",
    "\n",
    "print(django_avail.difference(struct_avail))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Check sequence queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VADLSKLSDS\n",
      "MADLSKLSDS\n",
      "EALGRELDGD*\n",
      "EALGRELDGD\n",
      "3546\n",
      "3545\n",
      "substrate prop, loading True\n",
      "VFPGQGAQWA\n",
      "VFPGQGAQWA\n",
      "AAVVPTLQRG\n",
      "AAVVPTLQRG\n",
      "294\n",
      "294\n",
      "domain\n",
      "AVVNGETAAL\n",
      "AVVNGETAAL\n",
      "PRALAEALAA\n",
      "PRALAEALAA\n",
      "66\n",
      "66\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "acc = 'BGC0000055.1'\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession=acc)\n",
    "subunit = cluster.subunits()[0]\n",
    "translated= Seq(subunit.getNucleotideSequence(), generic_dna).translate()\n",
    "# translated = Seq(subunit.getNucleotideSequence(), generic_dna).reverse_complement().translate()\n",
    "direct = subunit.getAminoAcidSequence()\n",
    "\n",
    "n = 10\n",
    "print(translated[:n])\n",
    "print(direct[:n])\n",
    "print(translated[len(translated)-n-1:])\n",
    "print(direct[len(direct)-n:])\n",
    "\n",
    "print(len(translated))\n",
    "print(len(direct))\n",
    "\n",
    "\n",
    "module = subunit.modules()[0]\n",
    "for domain in module.domains():\n",
    "    print(domain)\n",
    "    translated = Seq(domain.getNucleotideSequence(), generic_dna).translate()\n",
    "    # translated = Seq(domain.getNucleotideSequence(), generic_dna).reverse_complement().translate()\n",
    "    direct = domain.getAminoAcidSequence()\n",
    "\n",
    "    n = 10\n",
    "    print(translated[:n])\n",
    "    print(direct[:n])\n",
    "    print(translated[len(translated)-n:])\n",
    "    print(direct[len(direct)-n:])\n",
    "\n",
    "    print(len(translated))\n",
    "    print(len(direct))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Converting cluster to JSON file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'description': 'BE-14106',\n",
      " 'genbankAccession': 'FJ872523.1',\n",
      " 'mibigAccession': 'BGC0000029.1',\n",
      " 'architecture': {'becB': {0: {'AT': OrderedDict([('substrate', 'mal')]),\n",
      "                               'KR': {'active': False,\n",
      "                                      'type': 'U'}},\n",
      "                           1: {'AT': OrderedDict([('substrate', 'mmal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'B1'}}},\n",
      "                  'becD': {0: {'AT': OrderedDict([('substrate', 'mal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'C2'}},\n",
      "                           1: {'AT': OrderedDict([('substrate', 'mal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'U'}}},\n",
      "                  'becE': {0: {'AT': OrderedDict([('substrate', 'mal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'B1'}}},\n",
      "                  'becF': {0: {'AT': OrderedDict([('substrate', 'mmal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'B1'}},\n",
      "                           1: {'AT': OrderedDict([('substrate', 'mal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'B1'}}},\n",
      "                  'becG': {0: {'AT': OrderedDict([('substrate', 'mal')]),\n",
      "                               'KR': {'active': True,\n",
      "                                      'type': 'U'},\n",
      "                               'TE': OrderedDict([('cyclic', False)])}}}}\n"
     ]
    }
   ],
   "source": [
    "acc = 'BGC0000029.1'\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession=acc)\n",
    "pprint(pks.models.clusterDict(cluster))\n",
    "pks.models.clusterJSON('./corrections', cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pks.models.correctCluster('corrections/BGC0000029.1.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def correctCluster(filepath):\n",
    "    '''Function that takes as input a JSON file representing the changeable parameters\n",
    "       describing a PKS cluster and updates the database entry to reflect the information\n",
    "       contained in the JSON file.\n",
    "    '''\n",
    "    corr = json.loads(open(filepath).read(), object_pairs_hook=OrderedDict)\n",
    "    cluster = pks.models.Cluster.objects.get(mibigAccession=corr['mibigAccession'])\n",
    "    # Reorder subunits if necessary\n",
    "    newOrder = [str(x) for x in corr['architecture'].keys()]\n",
    "    cluster.reorderSubunits(newOrder)\n",
    "    # Change domain properties if necessary\n",
    "    for s,sdict in corr['architecture'].items():\n",
    "        for m,mdict in sdict.items():\n",
    "            for d,ddict in mdict.items():\n",
    "                if d == 'AT':\n",
    "                    pks.models.setSpecificity(cluster, s, int(m), d, ddict['substrate'])\n",
    "                elif d == 'KR':\n",
    "                    pks.models.setSpecificity(cluster, s, int(m), d, ddict['type'])\n",
    "                    pks.models.setActive(cluster, s, int(m), d, ddict['active'])\n",
    "                elif d in ['DH', 'ER']:\n",
    "                    pks.models.setActive(cluster, s, int(m), d, ddict['active'])\n",
    "                else:\n",
    "                    assert d == 'TE'\n",
    "                    cyclic = ddict['cyclic']\n",
    "    if cyclic:\n",
    "        pks.models.setCyclization(cluster, cyclic)\n",
    "\n",
    "correctCluster('corrections/BGC0000029.1.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[<rdkit.Chem.rdchem.Mol object at 0x7fe5b8154ba8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b713d588>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b713d5f8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b713d668>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b713d6d8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b713d748>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b713d7b8>,\n",
      " <rdkit.Chem.rdchem.Mol object at 0x7fe5b70ecdd8>]\n",
      "<class 'pks.models.KR'> False\n",
      "<class 'pks.models.KR'> True\n",
      "<class 'pks.models.KR'> True\n",
      "<class 'pks.models.KR'> True\n",
      "<class 'pks.models.KR'> True\n",
      "<class 'pks.models.KR'> True\n",
      "<class 'pks.models.KR'> True\n",
      "<class 'pks.models.KR'> True\n"
     ]
    }
   ],
   "source": [
    "acc = 'BGC0000029.1'\n",
    "cluster = pks.models.Cluster.objects.get(mibigAccession=acc)\n",
    "#pprint(cluster.architecture())\n",
    "pprint(cluster.computeProduct())\n",
    "\n",
    "for subunit in cluster.subunits():\n",
    "    for module in subunit.modules():\n",
    "        for domain in module.domains():\n",
    "            try:\n",
    "                print(type(domain), domain.active)\n",
    "            except:\n",
    "                pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
